src/backend/distributed/planner/distributed_planner.c

```diff
@@ -243,6 +254,113 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 }
 
 
+/*
+ * LocalReferenceTableSelect returns true if the query is a SELECT query that
+ * involves reference tables and can be executed locally.
+ */
+static bool
+IsLocalReferenceTableSelect(PlannerRestrictionContext *restrictionContext, Query *parse)
+{
+	bool currentNodeKnown = false;
+	bool allReferenceTables =
+		restrictionContext->relationRestrictionContext->allReferenceTables;
+
+	PrimaryNodeForGroup(GetLocalGroupId(), &currentNodeKnown);
+
+	/*
+	 * If there is an active primary entry for the current node in pg_dist_node,
+	 * then reference tables are also replicated to it.
+	 */
+
+	return allReferenceTables && currentNodeKnown && parse->commandType == CMD_SELECT;
+}
+
+
+/*
+ * PlanLocalReferenceTableSelect returns a plan for the given query that can be executed
+ * locally, assuming it satisfies the conditions in IsLocalReferenceTableSelect().
+ */
+static PlannedStmt *
+PlanLocalReferenceTableSelect(PlannerRestrictionContext *restrictionContext,
```

Comment:
Note that for fast-path router queries, we do not call `standard_planner()`, which means that we don't have a valid `restrictionContext `